[Compiler]: Using Located<T> instead of std::pair<SourceLoc, T>

kitaisreal opened the pull request:
1. Created Located<T> struct in swift/Basic
2. Updated usages throughout the compiler

https://bugs.swift.org/browse/SR-11889

@varungandhi-apple

kitaisreal committed:
SR-11889: Using Located<T> instead of std::pair<SourceLoc, T>

kitaisreal commented on include/swift/Basic/Located.h:
```diff
@@ -0,0 +1,39 @@
+  SourceLoc loc;
+
+  template<typename U>
+  friend bool operator ==(const Located<U> lhs, const Located<U> rhs) {
```
@varungandhi-apple There are a lot of places, where we use == with pair, should we remove this operator overloading and try to rewrite those places, without it ?

varungandhi-apple replied:
It's ok to have this overload. 👍 I did not realize that we were using `==` for `std::pair` but this makes sense.

varungandhi-apple replied:
Oh wait, could you mark the `lhs` and `rhs` as being passed by reference instead of by value? Don't want to accidentally copy the item in case it is a chonky string.

kitaisreal replied:
Sure. Fixed.

varungandhi-apple commented on include/swift/Basic/Located.h:
```diff
@@ -0,0 +1,39 @@
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2019 Apple Inc. and the Swift project authors
```
Sorry for nitpicking but this year should be just 2019.

kitaisreal replied:
Sure. Fixed.

varungandhi-apple commented on lib/AST/ASTContext.cpp:
```diff
@@ -1774,7 +1774,7 @@ ModuleDecl *ASTContext::getStdlibModule(bool loadIfAbsent) {
     auto mutableThis = const_cast<ASTContext*>(this);
     TheStdlibModule =
-      mutableThis->getModule({ std::make_pair(StdlibModuleName, SourceLoc()) });
+      mutableThis->getModule({{ StdlibModuleName, SourceLoc() }});
```
Usually, places in the compiler avoid two levels of `{{` because the types become less clear when reading the code, which is probably why we have that explicit `make_pair` there. Could you add an explicit constructor which initializes the fields and calls that instead? (There are 3-4 other places which are similar in having a `{{`)

kitaisreal replied:
Fixed.

varungandhi-apple commented on include/swift/Basic/Located.h:
```diff
@@ -0,0 +1,39 @@
+
+  T item;
+
+  SourceLoc loc;
```
Could you add some doc comments here for the fields and the type? For example [you don't need to stick to this wording if you have better ideas :)]

```
/// A currency type for keeping track of items which were found in the source code.
///
/// Several parts of the compiler need to keep track of a `SourceLoc` corresponding to an item, in case
/// they need to report some diagnostics later. For example, the ClangImporter needs to keep track
/// of where imports were originally written. Located makes it easy to do so while making the code
/// more readable, compared to using `std::pair`.
template<typename T>
 struct Located {

   /// The main item whose source location is being tracked.
   T item;

   /// The original source location from which the item was parsed.
   SourceLoc loc;
```

kitaisreal replied:
Sure. Updated with documentation.

varungandhi-apple commented on include/swift/Basic/Located.h:
```diff
@@ -0,0 +1,39 @@
+//===----------------------------------------------------------------------===//
+//
+// This file forward declares and imports various common LLVM datatypes that
+// swift wants to use unqualified.
```
This comment needs to be updated. Say something like

```
Provides a currency data type Located<T> that should be used instead of std::pair<T, SourceLoc>.
```

kitaisreal replied:
Fixed.

varungandhi-apple commented on include/swift/AST/ModuleLoader.h:
```diff
@@ -23,6 +23,7 @@
 #include "llvm/ADT/SetVector.h"
 #include "llvm/ADT/SmallSet.h"
 #include "llvm/ADT/TinyPtrVector.h"
+#include "swift/Basic/Located.h"
```
Could you order the imports so that `swift/Basic/Located.h` appears between `swift/Basic/LLVM.h` and `swift/Basic/SourceLoc.h`?

kitaisreal replied:
Yes. Fixed.

varungandhi-apple commented on include/swift/AST/ASTContext.h:
```diff
@@ -25,6 +25,7 @@
 #include "swift/AST/TypeAlignments.h"
 #include "swift/Basic/LangOptions.h"
 #include "swift/Basic/Malloc.h"
+#include "swift/Basic/Located.h"
```
Could you move the `#include` 1 line above, so that the `swift/Basic` section is lexicographically sorted?

kitaisreal replied:
Fixed.

varungandhi-apple commented on include/swift/Basic/Located.h:
```diff
@@ -0,0 +1,39 @@
+  friend bool operator ==(const Located<U> lhs, const Located<U> rhs) {
+    return lhs.item == rhs.item && lhs.loc == rhs.loc;
+  }
+};
```
Could you add a `dump` method that does something reasonable like for other types? You could have two overloads: `void dump() const; void dump(llvm::raw_ostream &OS) const` where the former calls the latter using `llvm::errs()`. [For example, `SourceFile` does this.]

You could add a new file `Located.cpp` which has the implementations and the appropriate `#includes`. You will also need to add `Located.cpp` to the `lib/Basic/CMakeLists.txt` file.

varungandhi-apple replied:
@kitaisreal what do you think about this point? I think it might be useful to have this method for debugging.

As to what is should print, perhaps something that prints the SourceLoc first and then the item, leaving a space in between (say)

kitaisreal replied:
@varungandhi-apple I agree, that we should introduce such function. Added.

varungandhi-apple commented:
Thanks for the PR! I've made some suggestions for changes. The `==` overload is fine, I just didn't think of it when writing the issue. Could you run a string search for `std::pair<SourceLoc,` (maybe you already did this and I missed it)? I think there are some places which have the type parameters the other way around and it would be to replace those usage sites too.

@swift-ci please smoke test

kitaisreal committed:
SR-11889: Fixed code review issues

kitaisreal commented:
Updated for pairs like std::pair<SourceLoc, ...>

varungandhi-apple commented on include/swift/Basic/Located.h:
```diff
@@ -2,16 +2,15 @@
 //
-// This file forward declares and imports various common LLVM datatypes that
-// swift wants to use unqualified.
+// Provides a currency data type Located<T> that should be used instead of std::pair<T, SourceLoc>.
```
Sorry, when I posted the doc comment example, I didn't check the line width. Could you wrap this doc comment and the later ones to < 80 chars?

kitaisreal replied:
Sure. Updated.

varungandhi-apple commented:
@swift-ci please smoke test

kitaisreal commented on include/swift/Basic/Located.h:
```diff
@@ -0,0 +1,50 @@
+  /// The original source location from which the item was parsed.
+  SourceLoc loc;
+
+  Located() {}
```
Here, I had to create default constructor because currently we are using these Located in DenseMap and there are code like
```c++

  value_type& FindAndConstruct(const KeyT &Key) {
    BucketT *TheBucket;
    if (LookupBucketFor(Key, TheBucket))
      return *TheBucket;

    return *InsertIntoBucket(TheBucket, Key);
  }

  ValueT &operator[](const KeyT &Key) {
    return FindAndConstruct(Key).second; /**Here**/
  }
```
/**Here**/ There is a call to default constructor. Should we make any workarounds around that @varungandhi-apple ?

varungandhi-apple replied:
Thanks for drawing attention to it, I missed that. Shouldn't the default constructor call the default constructors of the fields? My C++ initialization-fu isn't that great but I think it should instead be something like `Located() : item(), loc() {}` -- if we don't do that, I'm guessing we risk having the fields be uninitialized?

kitaisreal replied:
Yes, I agree this should be called otherwise it will be undefined behaviour. Thanks.

kitaisreal committed:
SR-11889: Fixed code review issues
1. Removed two braces {{ usage of Located initialization
2. Wrapped documentation into 80 characters

kitaisreal committed:
SR-11889: Fixed code review issues.

varungandhi-apple commented:
Sorry, I didn't pick this up earlier, but I think we should have the field names in `PascalCase`, so `Item` and `Loc`, for consistency with the rest of codebase.

varungandhi-apple commented on lib/FrontendTool/ImportedModules.cpp:
```diff
@@ -98,7 +98,7 @@ bool swift::emitImportedModules(ASTContext &Context, ModuleDecl *mainModule,
 
   if (opts.ImportUnderlyingModule) {
-    auto underlyingModule = clangImporter->loadModule(SourceLoc(), {{ mainModule->getName(), SourceLoc() }});
+    auto underlyingModule = clangImporter->loadModule(SourceLoc(), { Located<Identifier>(mainModule->getName(), SourceLoc()) });
```
Please wrap this line to < 80 columns.

kitaisreal replied:
Fixed.

kitaisreal commented:
@varungandhi-apple  Changed Located field names in PascalCase, Item and Loc.

kitaisreal committed:
SR-11889: Fixed code review issues
1. Updated Located field names with Pascal Case
2. Updated Located constuctor
3. Formatted lines with more than 80 symbols

varungandhi-apple commented:
@kitaisreal there are some merge conflicts. So you might want to rebase on top of latest `master`, fixing that conflict.

When you rebase, you'll pick up a patch I got merged recently (commit: https://github.com/apple/swift/pull/28479/commits/fb59293fe160556eb43e96fe70bb373b905237d7), which introduces a struct `Convention` -- that has two fields `StringRef ClangType` and `SourceLoc ClangTypeLoc`.  It would be better to represent it as one field `Located<StringRef> ClangType` instead and change downstream code appropriately. If you could adjust that as well, that'd be awesome. But yeah, not strictly necessary for the scope of your changes -- if you don't feel like doing it, feel free to skip that 😄.

kitaisreal committed:
SR-11889: Fixed code review issues
1. Use Located in Convention structure

kitaisreal commented:
@varungandhi-apple fixed conflicts, also updated Convention to use Located.

varungandhi-apple commented:
Awesome. I'm traveling for the next ~2 days, I'll get back to reviewing after that.

@swift-ci please smoke test

kitaisreal committed:
SR-11889: Located added dump methods

kitaisreal commented:
@varungandhi-apple currently the build from utils/build-script is passing without any errors. But on CI we are trying to build a toolchain and it is failing in lldb ? Can I reproduce such build error without trying to build a toolchain ?

varungandhi-apple commented:
You can use `build-script` to build LLDB as well. Here's one example:

```
./swift/utils/build-script \
  --lldb \
  --release-debuginfo \
  --test \
  -- \
  --skip-build-benchmarks \
  --lldb-use-system-debugserver \
  --skip-test-swift
```

CodaFi commented on include/swift/Basic/Located.h:
```diff
@@ -0,0 +1,57 @@
+  SWIFT_DEBUG_DUMP;
+  void dump(raw_ostream &os) const;
+
+  template<typename U>
```
Nit: LLVM style prefers the ref against the identifier, not the type.

varungandhi-apple replied:
This seems fine. For example, you can look at examples in this section: https://llvm.org/docs/CodingStandards.html#doxygen-use-in-documentation-comments

```
bool fooBar(bool Baz, StringRef Quux, std::vector<int> &Result);
```

varungandhi-apple replied:
Oh wait, I misread your comment. We are in agreement, but maybe your comment got moved to a different line or something, as there is nothing here.

kitaisreal replied:
Fixed.

CodaFi commented on include/swift/Basic/Located.h:
```diff
@@ -0,0 +1,57 @@
+
+  Located(): Item(), Loc() {}
+
+  Located(T Item, SourceLoc loc): Item(Item), Loc(loc) {}
```
Nit: Some whitespace is needed before the colons in both constructors.

kitaisreal replied:
Fixed.

CodaFi commented on include/swift/Basic/Located.h:
```diff
@@ -0,0 +1,57 @@
+  /// The original source location from which the item was parsed.
+  SourceLoc Loc;
+
+  Located(): Item(), Loc() {}
```
Hm, I wonder if it makes sense to have this constructor given that not every T can be default or value-initialized.  It might make more sense to provide an `explicit` constructor from a `T` or `const T&`, or default the `loc` in the other constructor and make it `explicit`.

CodaFi replied:
If this is the route you take, please also define an `explicit` move constructor.

varungandhi-apple replied:
I think the reason for adding this was that `DenseMap` required this constructor for its values?

CodaFi replied:
We should give a DenseMapInfo specialization to avoid that...

kitaisreal replied:
Hi, @CodaFi
As I understand correctly from code base we are making such specializations for Key:
```c++
namespace llvm {

template <typename T> struct DenseMapInfo;

template<typename T>
struct DenseMapInfo<swift::Located<T>> {

  static inline swift::Located<T> getEmptyKey() {
    return swift::Located<T>(T(), swift::SourceLoc());
  }

  static inline swift::Located<T> getTombstoneKey() {
    return swift::Located<T>(T(), swift::SourceLoc());
  }

  static unsigned getHashValue(const swift::Located<T> &LocatedVal) {
    return combineHashValue(DenseMapInfo<T>::getHashValue(LocatedVal.Item),
                            DenseMapInfo<swift::SourceLoc>::getHashValue(LocatedVal.Loc));
  }

  static bool isEqual(const swift::Located<T> &LHS, const swift::Located<T> &RHS) {
    return LHS == RHS;
  }
};
} // namespace llvm
```
But problem in DenseMap
```c++
value_type& FindAndConstruct(const KeyT &Key) {
    BucketT *TheBucket;
    if (LookupBucketFor(Key, TheBucket))
      return *TheBucket;

    return *InsertIntoBucket(TheBucket, Key);
  }

  ValueT &operator[](const KeyT &Key) {
    return FindAndConstruct(Key).second /**here**/;
  }
```
As I understand it tries to default initialize Value. And solution with DenseMapInfo specialization will not help in such case, where Located<T> is  Value. Is it right ?

CodaFi replied:
Tombstone and empty key requirements are not correctly implemented.  Take a look at how LLVM implements this for std::pair to get an idea of how this should look.  You need to delegate to the DenseMapInfo of T and the SourceLoc.

varungandhi-apple commented on include/swift/Basic/Located.h:
```diff
@@ -22,15 +21,27 @@
+
   template<typename U>
-  friend bool operator ==(const Located<U> lhs, const Located<U> rhs) {
+  friend bool operator ==(const Located<U>& lhs, const Located<U>& rhs) {
```
@CodaFi's comment was probably about this line, which should have `const Located<U> &lhs` instead of `const Located<U>& lhs` and similarly for `rhs`.

kitaisreal replied:
Fixed.

varungandhi-apple commented:
Mostly LGTM but you'll need to resolve the merge conflicts and address the failing LLDB compilation. Feel free to @ me or Robert (CodaFi) to run the tests.

kitaisreal committed:
SR-11889: Fixed code review issues

CodaFi reviewed (changes requested):
Github has marked my old comment as out of date, so I'm going to sticky this here:

Please provide a template specialization of `DenseMapInfo` that is modeled on `std::pair`'s DenseMapInfo, then remove the default constructor and follow the [rule of five](https://en.cppreference.com/w/cpp/language/rule_of_three).

kitaisreal committed:
SR-11889: Fixed code review issues

kitaisreal commented:
Hi, @varungandhi-apple, @CodaFi I added specialization with DenseMapInfo but It did not solve the problem with default constructor. Currently we have code in DenseMap 
```c++
  value_type& FindAndConstruct(const KeyT &Key) {
    BucketT *TheBucket;
    if (LookupBucketFor(Key, TheBucket))
      return *TheBucket;

    return *InsertIntoBucket(TheBucket, Key) /**HERE*//;
  }

  ValueT &operator[](const KeyT &Key) {
    return FindAndConstruct(Key).second;
  }
```
Here is code of InsertIntoBucket and we are calling it without any agruments.
```c++
template <typename KeyArg, typename... ValueArgs>
  BucketT *InsertIntoBucket(BucketT *TheBucket, KeyArg &&Key,
                            ValueArgs &&... Values) {
    TheBucket = InsertIntoBucketImpl(Key, Key, TheBucket);

    TheBucket->getFirst() = std::forward<KeyArg>(Key);
    ::new (&TheBucket->getSecond()) ValueT(std::forward<ValueArgs>(Values)...);
    return TheBucket;
  }
```
And if we follow rule of five, it will not help in such case.
I was able to make minimum reproduced example with https://godbolt.org/ compiler explorer, and LLVM library here https://godbolt.org/z/rrbpvh. It won't compile without providing a default constructor for Located.

kitaisreal commented:
@CodaFi, @varungandhi-apple could you please rerun tests ?

CodaFi commented:
@swift-ci please test

CodaFi commented:
@swift-ci please test

kitaisreal commented:
Hi, @CodaFi @varungandhi-apple I was able to sucessfully build lldb locally, but small changes were introduced in this repository https://github.com/apple/llvm-project. How should I push it ? Should I create pull request with the same name in https://github.com/apple/llvm-project. How will CI test two branches in two repositories together ?

CodaFi commented:
Branch from `swift/master` and submit your changes to `Apple/llvm-project` targeting `swift/master`.  We can trigger CI with paired pull requests and manually merge them as well once the holiday lock is over.

kitaisreal commented:
@CodaFi, @varungandhi-apple create pull request https://github.com/apple/llvm-project/pull/522.

varungandhi-apple commented:
apple/llvm-project#522

@swift-ci please test

varungandhi-apple commented:
@kitaisreal can you look at the test failures? It looks like `DiagnosticConsumerTests.cpp` is missing a `#include "swift/Basic/Located.h"`.

```
/Users/buildnode/jenkins/workspace/swift-PR-osx/branch-master/swift/unittests/AST/DiagnosticConsumerTests.cpp:20:30: error: no template named 'Located'
  using ExpectedDiagnostic = Located<StringRef>;
                             ^
```

kitaisreal committed:
SR-11889: Fixed code review issues

kitaisreal commented:
@varungandhi-apple Sure. Updated.

varungandhi-apple commented:
apple/llvm-project#522

@swift-ci please test

compnerd commented:
I think that this may have caused a build break for the Windows builds:

https://compnerd.visualstudio.com/swift-build/_build/results?buildId=23654&view=logs&j=3ee06e05-ce1c-5a30-d930-cdd7bd1c5236&t=c7be90c7-ae5e-5e2a-a2ec-6669b9e2ba36&l=3435

compnerd commented:
Additional breakage here: https://ci-external.swift.org/job/oss-swift-windows-x86_64/2363/

varungandhi-apple commented:
@compnerd yes, this is related. I merged it after the tests were passing for Linux and macOS. Is this an MSVC bug? Don't understand why it is complaining about a duplicate, there is exactly one `operator ==` present.

compnerd commented:
I dion't think that this is a bug at all.  I think that the compiler is identifying an issue with the code.  The warning is documented here https://docs.microsoft.com/en-us/cpp/error-messages/compiler-errors-2/compiler-error-c2995?view=vs-2019 .  The template is getting instantiated multiple times.  I think I have a fix for this, I'll push up a change.

compnerd commented:
See #29032 29032

varungandhi-apple commented:
I tried compiling this locally and I saw the same error with Clang. Wondering how it got through in CI. 😕

davezarzycki commented:
This fails to build on Fedora 31 (x86_64) and I'm surprised this builds at all. I can't find a declaration or definition of `operator<<(llvm::raw_ostream &, const SourceLoc &)` anywhere. How does this compile? Can we revert this? Or add the output operator? Or remove the offending line?

```
/home/dave/s/u/swift/lib/Basic/Located.cpp:23:6: error: invalid operands to binary expression ('llvm::raw_ostream' and 'const swift::SourceLoc')
  os << Loc << " " << Item;
  ~~ ^  ~~~
/home/dave/s/u/llvm/include/llvm/Support/raw_ostream.h:202:16: note: candidate function not viable: no known conversion from 'const swift::SourceLoc' to 'const void *' for 1st argument; take the address of the argument with &
  raw_ostream &operator<<(const void *P);
               ^
/home/dave/s/u/llvm/include/llvm/ADT/Optional.h:431:14: note: candidate function not viable: no known conversion from 'const swift::SourceLoc' to 'llvm::NoneType' for 2nd argument
raw_ostream &operator<<(raw_ostream &OS, NoneType);
             ^
/home/dave/s/u/llvm/include/llvm/Support/raw_ostream.h:146:16: note: candidate function not viable: no known conversion from 'const swift::SourceLoc' to 'char' for 1st argument
  raw_ostream &operator<<(char C) {
               ^
/home/dave/s/u/llvm/include/llvm/Support/raw_ostream.h:153:16: note: candidate function not viable: no known conversion from 'const swift::SourceLoc' to 'unsigned char' for 1st argument
  raw_ostream &operator<<(unsigned char C) {
               ^
/home/dave/s/u/llvm/include/llvm/Support/raw_ostream.h:160:16: note: candidate function not viable: no known conversion from 'const swift::SourceLoc' to 'signed char' for 1st argument
  raw_ostream &operator<<(signed char C) {
               ^
/home/dave/s/u/llvm/include/llvm/Support/raw_ostream.h:167:16: note: candidate function not viable: no known conversion from 'const swift::SourceLoc' to 'llvm::StringRef' for 1st argument
  raw_ostream &operator<<(StringRef Str) {
               ^
/home/dave/s/u/llvm/include/llvm/Support/raw_ostream.h:182:16: note: candidate function not viable: no known conversion from 'const swift::SourceLoc' to 'const char *' for 1st argument
  raw_ostream &operator<<(const char *Str) {
               ^
/home/dave/s/u/llvm/include/llvm/Support/raw_ostream.h:189:16: note: candidate function not viable: no known conversion from 'const swift::SourceLoc' to 'const std::string' (aka 'const basic_string<char, char_traits<char>, allocator<char> >') for 1st argument
  raw_ostream &operator<<(const std::string &Str) {
               ^
/home/dave/s/u/llvm/include/llvm/Support/raw_ostream.h:194:16: note: candidate function not viable: no known conversion from 'const swift::SourceLoc' to 'const SmallVectorImpl<char>' for 1st argument
  raw_ostream &operator<<(const SmallVectorImpl<char> &Str) {
               ^
/home/dave/s/u/llvm/include/llvm/Support/raw_ostream.h:198:16: note: candidate function not viable: no known conversion from 'const swift::SourceLoc' to 'unsigned long' for 1st argument
  raw_ostream &operator<<(unsigned long N);
               ^
/home/dave/s/u/llvm/include/llvm/Support/raw_ostream.h:199:16: note: candidate function not viable: no known conversion from 'const swift::SourceLoc' to 'long' for 1st argument
  raw_ostream &operator<<(long N);
               ^
/home/dave/s/u/llvm/include/llvm/Support/raw_ostream.h:200:16: note: candidate function not viable: no known conversion from 'const swift::SourceLoc' to 'unsigned long long' for 1st argument
  raw_ostream &operator<<(unsigned long long N);
               ^
/home/dave/s/u/llvm/include/llvm/Support/raw_ostream.h:201:16: note: candidate function not viable: no known conversion from 'const swift::SourceLoc' to 'long long' for 1st argument
  raw_ostream &operator<<(long long N);
               ^
/home/dave/s/u/llvm/include/llvm/Support/raw_ostream.h:204:16: note: candidate function not viable: no known conversion from 'const swift::SourceLoc' to 'unsigned int' for 1st argument
  raw_ostream &operator<<(unsigned int N) {
               ^
/home/dave/s/u/llvm/include/llvm/Support/raw_ostream.h:208:16: note: candidate function not viable: no known conversion from 'const swift::SourceLoc' to 'int' for 1st argument
  raw_ostream &operator<<(int N) {
               ^
/home/dave/s/u/llvm/include/llvm/Support/raw_ostream.h:212:16: note: candidate function not viable: no known conversion from 'const swift::SourceLoc' to 'double' for 1st argument
  raw_ostream &operator<<(double N);
               ^
/home/dave/s/u/llvm/include/llvm/Support/raw_ostream.h:229:16: note: candidate function not viable: no known conversion from 'const swift::SourceLoc' to 'const llvm::format_object_base' for 1st argument
  raw_ostream &operator<<(const format_object_base &Fmt);
               ^
/home/dave/s/u/llvm/include/llvm/Support/raw_ostream.h:232:16: note: candidate function not viable: no known conversion from 'const swift::SourceLoc' to 'const llvm::FormattedString' for 1st argument
  raw_ostream &operator<<(const FormattedString &);
               ^
/home/dave/s/u/llvm/include/llvm/Support/raw_ostream.h:235:16: note: candidate function not viable: no known conversion from 'const swift::SourceLoc' to 'const llvm::FormattedNumber' for 1st argument
  raw_ostream &operator<<(const FormattedNumber &);
               ^
/home/dave/s/u/llvm/include/llvm/Support/raw_ostream.h:238:16: note: candidate function not viable: no known conversion from 'const swift::SourceLoc' to 'const llvm::formatv_object_base' for 1st argument
  raw_ostream &operator<<(const formatv_object_base &);
               ^
/home/dave/s/u/llvm/include/llvm/Support/raw_ostream.h:241:16: note: candidate function not viable: no known conversion from 'const swift::SourceLoc' to 'const llvm::FormattedBytes' for 1st argument
  raw_ostream &operator<<(const FormattedBytes &);
               ^
/home/dave/s/u/llvm/include/llvm/ADT/Optional.h:435:14: note: candidate template ignored: could not match 'Optional<type-parameter-0-0>' against 'const swift::SourceLoc'
raw_ostream &operator<<(raw_ostream &OS, const Optional<T> &O) {
             ^
1 error generated.
```

davezarzycki commented:
There are two fundamental problems here:

1) This code doesn't compile with top-of-tree clang, which is more reliable about emitting templated functions that are marked "used", i.e. what `SWIFT_DEBUG_DUMP` does among other things. This same bug was tripped over when the SWIFT_DEBUG_DUMP macro was introduced.
2) `operator<<` on a SourceLoc is arguably meaningless. One also needs a `SourceManager` reference and a buffer ID in order to extract the line and column information.

I'm just going to comment the problematic line out, since older clangs don't emit the function anyway and @varungandhi-apple seemingly did not notice. I'll update this comment with a pull request reference soon.
